internal/daemon/daemon.go

```diff
@@ -622,11 +652,87 @@ func (d *Daemon) StartAPI(bootstrap bool, initConfig map[string]string, newConfi
 	return nil
 }
 
+// UpdateServers updates and start/stops the additional listeners.
+func (d *Daemon) UpdateServers() error {
+	var activeExtensionServers []string
+	for name := range d.config.GetServers() {
+		activeExtensionServers = append(activeExtensionServers, name)
+	}
+
+	// Stop all additional listeners which got removed from the config.
+	for _, serverName := range d.ExtensionServers() {
+		if !shared.ValueInSlice(serverName, activeExtensionServers) {
+			// Remove their config.
+			d.extensionServersMu.Lock()
+			extensionServer, ok := d.extensionServers[serverName]
+			if !ok {
+				// There isn't any additional listener set that matches this name.
+				d.extensionServersMu.Unlock()
+				continue
+			}
+
+			// Set an empty config for this specific additional listener.
+			extensionServer.ServerConfig = types.ServerConfig{}
+
+			// Reassign the map struct.
+			d.extensionServers[serverName] = extensionServer
+			d.extensionServersMu.Unlock()
+
+			err := d.endpoints.DownByName(serverName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Restart all additional listeners which received an update.
+	for serverName, extensionServerConfig := range d.config.GetServers() {
+		// Lock the entire section to ensure the reassignment can happen.
+		d.extensionServersMu.Lock()
+		extensionServer, ok := d.extensionServers[serverName]
+		if !ok {
+			// There isn't any additional listener set that matches this name.
+			d.extensionServersMu.Unlock()
+			continue
+		}
+
+		modified := false
+		if extensionServer.ServerConfig != extensionServerConfig {
+			modified = true
+		}
+
+		extensionServer.ServerConfig = extensionServerConfig
+
+		// Reassign the map struct.
+		d.extensionServers[serverName] = extensionServer
+		d.extensionServersMu.Unlock()
+
+		// Stop the additional listener in case it got modified.
+		if modified {
+			err := d.endpoints.DownByName(serverName)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Start any additional listener.
+	// This operation is idempotent.
+	err := d.addExtensionServers(false, d.ClusterCert(), d.Address().URL.Host)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
 // startUnixServer starts up the core unix listener with the given resources.
 func (d *Daemon) startUnixServer(serverEndpoints []rest.Resources) error {
 	ctlServer := d.initServer(serverEndpoints...)
 	ctl := endpoints.NewSocket(d.shutdownCtx, ctlServer, d.os.ControlSocket(), d.os.SocketGroup)
-	d.endpoints = endpoints.NewEndpoints(d.shutdownCtx, ctl)
+	d.endpoints = endpoints.NewEndpoints(d.shutdownCtx, map[string]endpoints.Endpoint{
+		"unix": ctl,
```

Comment:
Does this and `core` below deserve a constant?

Reply:
Fixed